pkg/csi/cinder/openstack/openstack_backups.go

```diff
@@ -0,0 +1,195 @@
+	return backup, nil
+}
+
+// ListBackups retrieves a list of active backups from Cinder for the corresponding Tenant.  We also
+// provide the ability to provide limit and offset to enable the consumer to provide accurate pagination.
+// In addition the filters argument provides a mechanism for passing in valid filter strings to the list
+// operation.  Valid filter keys are:  Name, Status, VolumeID, Limit, Marker (TenantID has no effect).
+func (os *OpenStack) ListBackups(filters map[string]string) ([]backups.Backup, error) {
+	var allBackups []backups.Backup
+
+	// Build the Opts
+	opts := backups.ListOpts{}
+	for key, val := range filters {
+		switch key {
+		case "Status":
+			opts.Status = val
+		case "Name":
+			opts.Name = val
+		case "VolumeID":
+			opts.VolumeID = val
+		case "Marker":
+			opts.Marker = val
+		case "Limit":
+			opts.Limit, _ = strconv.Atoi(val)
+		default:
+			klog.V(3).Infof("Not a valid filter key %s", key)
+		}
+	}
+	mc := metrics.NewMetricContext("backup", "list")
+
+	allPages, err := backups.List(os.blockstorage, opts).AllPages()
+	if err != nil {
+		panic(err)
+	}
+	allBackups, err = backups.ExtractBackups(allPages)
+	if err != nil {
+		panic(err)
+	}
+
+	if mc.ObserveRequest(err) != nil {
+		return nil, err
+	}
+
+	return allBackups, nil
+}
+
+// DeleteBackup issues a request to delete the Backup with the specified ID from the Cinder backend.
+func (os *OpenStack) DeleteBackup(backupID string) error {
+	mc := metrics.NewMetricContext("backup", "delete")
+	err := backups.Delete(os.blockstorage, backupID).ExtractErr()
+	if mc.ObserveRequest(err) != nil {
+		klog.Errorf("Failed to delete backup: %v", err)
+	}
+	return err
+}
+
+// GetBackupByID returns backup details by id.
+func (os *OpenStack) GetBackupByID(backupID string) (*backups.Backup, error) {
+	mc := metrics.NewMetricContext("backup", "get")
+	backup, err := backups.Get(os.blockstorage, backupID).Extract()
+	if mc.ObserveRequest(err) != nil {
+		klog.Errorf("Failed to get backup: %v", err)
+		return nil, err
+	}
+	return backup, nil
+}
+
+// WaitBackupReady waits until backup is ready. It waits longer depending on
+// the size of the corresponding snapshot.
+func (os *OpenStack) WaitBackupReady(backupID string, snapshotSize int, backupMaxDurationSecondsPerGB int) error {
+	var err error
+
+	duration := time.Duration(backupMaxDurationSecondsPerGB*snapshotSize + 30)
+	ctx := context.Background()
+
+	err = os.waitBackupReadyWithContext(ctx, backupID, duration)
+	if wait.Interrupted(err) {
+		err = fmt.Errorf("timeout, Backup %s is still not Ready %v", backupID, err)
+	}
+
+	return err
+}
+
+// Supporting function for WaitBackupReady().
+// Allows for a timeout while waiting for the backup to be ready.
+func (os *OpenStack) waitBackupReadyWithContext(ctx context.Context, backupID string, duration time.Duration) error {
+	_, cancel := context.WithTimeout(context.Background(), duration*time.Second)
+	defer cancel()
+	var ok bool
+	var err error
+	for !ok {
```

Comment:
It's better to use context's `Done` channel, see https://pkg.go.dev/context#WithTimeout

```go
	ctx, cancel := context.WithTimeout(context.Background(), duration)
	defer cancel()

	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			ok, err = os.backupIsReady(backupID)
			if err != nil {
				return err
			}

			if ok {
				return nil
			}
		case <-ctx.Done():
			return ctx.Err()
		}
	}

	return nil
```

Reply:
Thank you, this looks much better than what I made up